File: src/main/java/com/qlm/zombie/dayphase/DayPhaseManager.java
package com.qlm.zombie.dayphase;

import com.mojang.logging.LogUtils;
import com.qlm.zombie.config.QLMConfig;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.Difficulty;
import net.minecraft.world.level.Level;
import net.minecraftforge.event.TickEvent;
import net.minecraftforge.event.server.ServerStartedEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.server.ServerLifecycleHooks;
import org.slf4j.Logger;

public class DayPhaseManager {

    private static final Logger LOGGER = LogUtils.getLogger();
    private static final long DAY_LENGTH = 57600L;  // 2400 tick = 1 小时
    private static int tickCounter = 0;
    private static boolean firstDayInitialized = false;

    @SubscribeEvent
    public static void onServerStarted(ServerStartedEvent event) {
        MinecraftServer server = event.getServer();
        ServerLevel overworld = server.getLevel(Level.OVERWORLD);
        if (overworld == null) return;

        long currentDay = overworld.getDayTime() / DAY_LENGTH;
        DayPhase phase = DayPhase.forDay(currentDay);

        Difficulty currentDifficulty = overworld.getDifficulty();
        Difficulty targetDifficulty = phase.difficulty();

        if (currentDifficulty != targetDifficulty) {
            server.setDifficulty(targetDifficulty, true);
            LOGGER.info("[QLM Zombie] 服务器启动: 第{}天 -> 设为阶段 {} (难度 {})", currentDay, phase.displayName(), targetDifficulty.getKey());
        }

        DifficultyLockState lockState = DifficultyLockState.get(server);
        lockState.setLastAppliedDay(currentDay);
        lockState.setLastPhase(phase.name());
        lockState.setLocked(phase.isLocked());
        firstDayInitialized = true;
    }

    @SubscribeEvent
    public static void onServerTick(TickEvent.ServerTickEvent event) {
        if (event.phase != TickEvent.Phase.END) return;
        MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
        if (server == null) return;
        ServerLevel overworld = server.getLevel(Level.OVERWORLD);
        if (overworld == null) return;

        if (++tickCounter < 200) return;
        tickCounter = 0;

        long dayTime = overworld.getDayTime();
        long currentDay = dayTime / DAY_LENGTH;

        DayPhase phase = DayPhase.forDay(currentDay);

        DifficultyLockState lockState = DifficultyLockState.get(server);

        if (!firstDayInitialized && currentDay >= 0) {
            firstDayInitialized = true;
            Difficulty initialDifficulty = overworld.getDifficulty();

            if (initialDifficulty != phase.difficulty()) {
                forceSetDifficulty(server, phase.difficulty());
                LOGGER.info("[QLM Zombie] 游戏初始化 -> 第{}天 -> 阶段 {} (难度 {})", currentDay, phase.displayName(), phase.difficulty().getKey());
            }
            lockState.setLastAppliedDay(currentDay);
            lockState.setLastPhase(phase.name());
            return;
        }

        if (phase.isLocked()) {
            lockState.setLocked(true);
        } else {
            lockState.setLocked(false);
        }

        if (QLMConfig.ENABLE_DIFFICULTY_LOCK.get() && lockState.isLocked()) {
            if (overworld.getDifficulty() != Difficulty.HARD) {
                forceSetDifficulty(server, Difficulty.HARD);
                LOGGER.info("[QLM Zombie] 难度已锁定为 HARD (第{}天，{}阶段)，禁止更改", currentDay, phase.displayName());
            }
        } else if (lockState.getLastAppliedDay() != currentDay) {
            if (overworld.getDifficulty() != phase.difficulty()) {
                forceSetDifficulty(server, phase.difficulty());
                LOGGER.info("[QLM Zombie] 第{}天 -> 阶段 {} (难度 {})", currentDay, phase.displayName(), phase.difficulty().getKey());
            }
            lockState.setLastAppliedDay(currentDay);
            lockState.setLastPhase(phase.name());
        }
    }

    private static void forceSetDifficulty(MinecraftServer server, Difficulty difficulty) {
        server.setDifficulty(difficulty, true);
    }

    public static DayPhase getCurrentPhase() {
        MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
        if (server == null) return DayPhase.SAFE;

        ServerLevel overworld = server.getLevel(Level.OVERWORLD);
        if (overworld == null) return DayPhase.SAFE;

        long dayTime = overworld.getDayTime();
        long currentDay = dayTime / DAY_LENGTH;
        return DayPhase.forDay(currentDay);
    }

    public static long getCurrentDay() {
        MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
        if (server == null) return 0;

        ServerLevel overworld = server.getLevel(Level.OVERWORLD);
        if (overworld == null) return 0;

        return overworld.getDayTime() / DAY_LENGTH;
    }
}

File: src/main/kotlin/com/qlm/zombie/dayphase/DayPhaseManager.kt
package com.qlm.zombie.dayphase

import com.qlm.zombie.QLMZombieMod
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.level.Level
import net.minecraftforge.event.TickEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.api.distmarker.Dist

@Mod.EventBusSubscriber(modid = QLMZombieMod.MOD_ID, bus = Mod.EventBusSubscriber.Bus.FORGE, value = [Dist.DEDICATED_SERVER])
object DayPhaseManager {
    private var currentPhase: DayPhase = DayPhase.PEACE
    private var lastDay: Int = -1
    private var lockApplied: Boolean = false

    @SubscribeEvent
    fun onLevelTick(event: TickEvent.LevelTickEvent) {
        if (event.phase != TickEvent.Phase.END) return
        if (event.level.isClientSide) return
        if (event.level.dimension() != Level.OVERWORLD) return

        val serverLevel = event.level as? ServerLevel ?: return
        val day = (serverLevel.dayTime / 24000L).toInt()

        if (day != lastDay) {
            lastDay = day
            val newPhase = DayPhase.fromDay(day)
            if (newPhase != currentPhase) {
                val oldPhase = currentPhase
                currentPhase = newPhase
                onPhaseChange(oldPhase, newPhase, serverLevel)
            }
        }

        // 100 天+：锁定困难，强制难度为困难且无法更改
        if (currentPhase == DayPhase.LOCKED_HARD && !lockApplied) {
            lockApplied = true
            applyHardLock(serverLevel)
        }
    }

    private fun applyHardLock(level: ServerLevel) {
        try {
            // 持久化锁定状态
            val state = DifficultyLockState.get(level.server)
            state.setLocked(true)
            state.setLastAppliedDay(getCurrentDay())
            state.setLastPhase(DayPhase.LOCKED_HARD.name)
            // 强制困难难度
            if (level.difficulty != net.minecraft.world.Difficulty.HARD) {
                level.server.setDifficulty(net.minecraft.world.Difficulty.HARD, true)
            }
            for (player in level.players()) {
                player.sendSystemMessage(
                    Component.literal("§4[昼夜阶段] §c服务器已进入 §6锁定困难 §c阶段（第 100 天+），难度锁定为困难，无法更改！")
                )
            }
            QLMZombieMod.LOGGER.info("[昼夜阶段] 已锁定困难难度（第 {} 天后）", getCurrentDay())
        } catch (e: Exception) {
            QLMZombieMod.LOGGER.warn("[昼夜阶段] 应用困难锁定失败: {}", e.message)
        }
    }

    /** 难度是否已锁定（100 天+） */
    @JvmStatic
    fun isDifficultyLocked(): Boolean = currentPhase == DayPhase.LOCKED_HARD

    /** 阻止在锁定阶段更改难度（供命令等调用，返回是否被阻止） */
    @JvmStatic
    fun blockDifficultyChange(): Boolean {
        if (isDifficultyLocked()) return true
        val server = net.minecraftforge.server.ServerLifecycleHooks.getCurrentServer() ?: return false
        val state = DifficultyLockState.get(server)
        return state.isLocked()
    }

    private fun onPhaseChange(oldPhase: DayPhase, newPhase: DayPhase, level: ServerLevel) {
        val msg = "§6[昼夜阶段] §e从 ${oldPhase.name} 切换到 ${newPhase.name} §7(难度乘数: ${newPhase.difficultyMultiplier}x)" +
            if (newPhase.isLocked()) " §c[难度已锁定，无法更改]" else ""
        for (player in level.players()) {
            player.sendSystemMessage(Component.literal(msg))
        }
        QLMZombieMod.LOGGER.info(
            "[昼夜阶段] 阶段变更: {} -> {}, 难度乘数: {}x",
            oldPhase.name,
            newPhase.name,
            newPhase.difficultyMultiplier
        )
    }

    @JvmStatic
    fun getCurrentPhase(): DayPhase = currentPhase

    @JvmStatic
    fun getCurrentDay(): Long {
        val server = net.minecraftforge.server.ServerLifecycleHooks.getCurrentServer() ?: return 0L
        val overworld = server.getLevel(Level.OVERWORLD) ?: return 0L
        return overworld.dayTime / 24000L
    }

    fun reset() {
        currentPhase = DayPhase.PEACE
        lastDay = -1
        lockApplied = false
    }
}